source/nodejs/adaptivecards/src/markdown-formatter.ts

```diff
@@ -1,182 +1,186 @@
+                end: "</ul>"
+            },
+            patterns: ['-', '\\r|\\n'],
+            patternType: 'list',
+            groups: 1,
+        }, {
+            type: "bold",
+            styles: {
+                start: "<b>",
+                end: "</b>"
+            },
+            patterns: ['**'],
+            patternType: 'symmetric',
             groups: 1,
-        },  
-		{
-			type: 'bold',
-            styles:{
-            start: "<b>",
-            end: "</b>"},
-			pattern: ['**'],
-			patternType: 'symmetric',
-			groups: 1,
-		},
-		{
-			type: 'italic',
-            styles:{
+        }, {
+            type: "italic",
+            styles: {
                 start: "<i>",
-                end: "</i>"},
-			pattern: ['_'],
-			patternType: 'symmetric',
-			groups: 1,
-		},
-		{
-			type: 'hyperlink',
+                end: "</i>"
+            },
+            patterns: ['_'],
+            patternType: 'symmetric',
+            groups: 1,
+        }, {
+            type: "hyperlink",
             styles: {
-                html:"<a href='{1}'>{0}</a>"
+                html: "<a href='{1}'>{0}</a>"
             },
-			pattern: ['[]()'],
-			patternType: 'asymmetric',
-			groups: 2,
-		},
+            patterns: ['[]()'],
+            patternType: 'asymmetric',
+            groups: 2,
+        }
     ];
 
-    //combines default config and user config.
-    static setCustomMarkdownRegex(regexArray: any[]){
-        MarkdownFormatter.regexArray = regexArray;
-
-		// prefer user configs
-		for(var i = 0; i < MarkdownFormatter.MD_FORMATTER_CONFIG.length; i++){
-			for(var j =0; j < MarkdownFormatter.regexArray.length; j++){
-				if(MarkdownFormatter.MD_FORMATTER_CONFIG[i].type == MarkdownFormatter.regexArray[j].type){
-					MarkdownFormatter.MD_FORMATTER_CONFIG[i] = MarkdownFormatter.regexArray[j];
-					MarkdownFormatter.regexArray.splice(j, 1);
-					continue;					
-				}
-			}
-		}
-        MarkdownFormatter.regexArray = MarkdownFormatter.MD_FORMATTER_CONFIG.concat(MarkdownFormatter.regexArray);
-        
+    constructor() {
+        super();
+        this.configuration.forEach(config => {
+            this.initailize(config);
+        });
     }
-    
-
-	constructor(text:string) {
-        this.text = text;
-       
-		for (var i = 0; i < MarkdownFormatter.MD_FORMATTER_CONFIG.length; i++) {
-			let patternArray = MarkdownFormatter.MD_FORMATTER_CONFIG[i].pattern;
-			let patternType = MarkdownFormatter.MD_FORMATTER_CONFIG[i].patternType;
-            let groups = MarkdownFormatter.MD_FORMATTER_CONFIG[i].groups;
-            let pattern = patternArray[0];
-            
-            if(patternType === "custom"){                
-                pattern = patternArray[0];
+
+    /**
+     * To add custom markdown rules. This will update the default markdown rule is the same type is passed.
+     * 
+     * @param configurations 
+     */
+    public setCustomRules(configurations: IFormatterConfiguration[]) {
```

Comment:
I still believe this isn't necessary or ideal. Please refer to my main comment.

Reply:
When a user adds his own custom regex we need to call the initialize method to update the patterns list. I tried to make it a part of constructor but the constructor in the card-elements gets called even before the user sets custom rules. How else to do?

Reply:
How about this: design configuration as I suggest, keep the `MarkdownFormatter._configuration` field private, allow the developer to pass a custom configuration to the constructor and remove `setCustomRules`. This way you can initialize everything just once at construction time, and if the developer wants to use multiple different configurations they can simply instantiate multiple instances of `MarkdownFormatter`.

Reply:
Then I would have to expose MarkdownFormatter class as a static variable right? I did the following changes to make it work. 

I added this is card-elements.ts
```
static MarkdownFormatter = MarkdownFormatter;
```
This is how my a user can pass the custom rules.
```
AdaptiveCards.AdaptiveCard.textFormatter = new AdaptiveCards.AdaptiveCard.MarkdownFormatter(customRules);
```

Am I doing it right?

Reply:
Here is what I am suggesting:
```typescript
export type FormatterPatternType = "start-end" | "symmetric" | "asymmetric";

export interface IFormatterRule {
    styles: {
        start?: string;
        end?: string;
        html?: string;
    };
    patterns: string[];
    patternType: FormatterPatternType;
    groups: number;
}

export interface IFormatterConfiguration {
    orderedList: IFromatterRule;
    unorderedList: IFormatterRule;
    bold: IFormatterRule;
    italic: IFormatterRule;
    hyperlink: IFormatterRule;
    [name: string]: IFormatterRule;
}

export class MarkdownFormatter {
    static getDefaultConfiguration(): IFormatterConfiguration {
        return {
            orderedList: {
                styles:{
                    start: "<ol>",
                    end: "</ol>"
                },
                patterns: ["\\d.", '\\r|\\n'],
                patternType: "start-end",
                groups: 1
            },
            unorderedList: {
                ...
            },
            ...
        }
    }

    private _configuration: IFormatterConfiguration;

    constructor(configuration?: IFormatterConfiguration) {
        super();

        this._configuration = configuration ? configuration : MarkdownFormatter.getDefaultConfiguration();

        // Initialize configuration
    }
}
```

No need to have a `static MarkdownFormatter`. Developers can write this:
```typescript
// This instantiates a MarkdownFormatter with the default configuration
AdaptiveCards.AdaptiveCard.textFormatter = new AdaptiveCards.AdaptiveCard.MarkdownFormatter();

// The below instantiates a MarkdownFormatter with a custom configuration

// Get a copy of the default configuration
let myConfiguration = MarkdownFormatter.getDefaultCOnfiguration();

// Customize the orderedList rule
myConfiguration.orderedList = {
    ...
}

// Add a custom rule
myConfiguration["myCustomRule"] = {
    patternType: "start-end",
    ...
};

// Alternatively, the custom configuration can be created entirely from scratch
myConfiguration = {
    orderedList: {
        ...
    },
    unorderedList: {
        ...
    },
    ...,
    myCustomRule: {
        ...
    }
};

// Pass the custom configuration to the MarkdownFormatter constructor
AdaptiveCards.AdaptiveCard.textFormatter = new AdaptiveCards.AdaptiveCard.MarkdownFormatter(myConfiguration);

```
The point for all this really is that everything is strongly typed, and thus the developer gets "IntelliSense" hints as they code.

Does that make sense?

Reply:
@dclaux  Yes. This is brilliant. Thanks so much for the detailed explanation. I have made these changes. Can you please review it.